[FLINK-13339][ml] Add an implementation of pipeline's api

xuyang1706 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Add an implementation of pipeline's api


## Brief change log

  - *Add an implement PipelineStage, Estimator, Transformer, Model.*
  - *Add MLSession to hold the execution environment and others session shared variable.*
  - *Add AlgoOperator for the implementation of algorithms.*
  - *Add BatchOperator and StreamOperator based on AlgoOperator*
  - *Add TableSourceBatchOp and TableSourceStreamOp *


## Verifying this change
This change added tests and can be verified as follows:

- run test case pass

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLSession.java:
```diff
@@ -0,0 +1,187 @@
+/**
+ * MLSession hold the execution environment and others session shared variable.
+ *
+ * <P>In machine learning, we find that a single execution context is convenient for the user
```
What is the scope of an MLSession? Do we allow multiple MLSessions in the same JVM? If so, how are we going to support that?

xuyang1706 replied:
we defined MLEnvironmentFactory to replace the global static MLSession and support to set multiple Environment

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/Model.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @param <M> The class type of the {@link Model} implementation itself
+ */
+public abstract class Model<M extends Model <M>> extends Transformer <M>
```
Same here, can we change the class name to `ModelBase`?

xuyang1706 replied:
thx, changed

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/PipelineStage.java:
```diff
@@ -0,0 +1,67 @@
+ * @param <S> The class type of the {@link PipelineStage} implementation itself, used by {@link
+ *            org.apache.flink.ml.api.misc.param.WithParams} and Cloneable.
+ */
+public abstract class PipelineStage<S extends PipelineStage <S>>
```
Same here, `PipelineStageBase`? This class does not implement the `org.apache.flink.ml.api.core.PipelineStage`, is this intended?

walterddr replied:
I am not sure this class is a must. I think each type of pipeline stage individually can set its own base class. see: https://stackoverflow.com/a/3599379/11332462

xuyang1706 replied:
> Same here, `PipelineStageBase`? This class does not implement the `org.apache.flink.ml.api.core.PipelineStage`, is this intended?

> I am not sure this class is a must. I think each type of pipeline stage individually can set its own base class. see: https://stackoverflow.com/a/3599379/11332462

It has been renamed and PipelineStage is the standard concept in pipeline and if it implements the `WithParams` , the subclasses could do not care about `WithParams`

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/Transformer.java:
```diff
@@ -0,0 +1,100 @@
+ * @param <T> The class type of the {@link Transformer} implementation itself, used by {@link
+ *            org.apache.flink.ml.api.misc.param.WithParams}
+ */
+public abstract class Transformer<T extends Transformer <T>>
```
Same here, `TransformerBase`?

xuyang1706 replied:
thx, changed

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/streamoperator/StreamOperator.java:
```diff
@@ -0,0 +1,86 @@
+/**
+ * Base class of streaming algorithm operators.
+ */
+public abstract class StreamOperator<T extends StreamOperator <T>> extends AlgoOperator<T> {
```
The name `StreamOperator` is colliding with `org.apache.flink.streaming.api.operators.StreamOperator` which is a widely established class in Flink. Can we change it to something like `StreamAlgoOperator`? We may need to change the `BatchOperator` to `BatchAlgoOperator` as well.

xuyang1706 replied:
Thanks for your advice and offline discussion. In our ML algorithm implementations and applications, we need not use org.apache.flink.streaming.api.operators.StreamOperator, and we'd like to use BatchOperator in ML lib.

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/streamoperator/StreamOperator.java:
```diff
@@ -0,0 +1,86 @@
+		return getOutput().toString();
+	}
+
+	public <S extends StreamOperator> S link(S next) {
```
I am curious why `link()` and `linkTo()` methods require a template type while `linkFrom()` does not need that?

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/streamoperator/StreamOperator.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.List;
+
+/**
+ * Base class of streaming algorithm operators.
```
This class seems needs some java doc for the public methods.

xuyang1706 replied:
thx, added

ex00 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,86 @@
+		if (null != ins && ins.size() == 1) {
+			return linkFrom(ins.get(0));
+		} else {
+			throw new RuntimeException("Not support more than 1 inputs!");
```
Implementations of `linkFrom(BatchOperator, BatchOperator)`, `linkFrom(BatchOperator, BatchOperator,BatchOperator)`, `linkFrom(List)` looks inconsistent.
In methods for BatchOperators is created list of elements more that 1. but in result is throw exception that count elements should be 1.

ex00 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	public T linkFrom(List <BatchOperator> ins) {
+		if (null != ins && ins.size() == 1) {
```
what is difference on this case from `linkFrom(BatchOperator)`?

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,80 @@
+		return next;
+	}
+
+	public abstract T linkFrom(BatchOperator<?>... inputs);
```
This is a new abstract API. Please provide JavaDoc

xuyang1706 replied:
Thanks, added intro and example.
For there is no answer area of the above question, I put the answer here:
PipelineStage only supports single input and single output, it is the basic unit for pipeline. AlgoOperator supports multi-input and multi-output. We’d like to implement the algorithm with  AlgoOperator, and PipelineStage’s fit and transform function can call the AlgoOperator.

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,80 @@
+/**
+ * Base class of batch algorithm operators.
+ */
+public abstract class BatchOperator<T extends BatchOperator<T>> extends AlgoOperator<T> {
```
please provide java doc for `link` `linkTo` and `linkFrom`.

xuyang1706 replied:
Thanks, provided.

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,80 @@
+	}
+
+	public <B extends BatchOperator<?>> B linkTo(B next) {
+		next.linkFrom(this);
```
what's the rational of making link in this direction? what if downstream operator calls `linkFrom` twice with the same upstream; or twice with different upstreams?

xuyang1706 replied:
> what's the rational of making link in this direction? what if downstream operator calls `linkFrom` twice with the same upstream; or twice with different upstreams?

thx

1. with this method, the operator can link from source to sink if every operator in the dag is single input and single output and it is straightforward that execute in order from front to back
2. It is not recommended to linkFrom itself or link the same group inputs twice（added to javadoc） and the implement of the operator will define the behavior

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/source/TableSourceBatchOp.java:
```diff
@@ -0,0 +1,44 @@
+
+	@Override
+	public TableSourceBatchOp linkFrom(BatchOperator<?>... inputs) {
+		throw new UnsupportedOperationException("Not supported.");
```
more concise error message: `"Table source operator should not have any upstream to link from"`

xuyang1706 replied:
Thanks, this message is more concise.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/Estimator.java:
```diff
@@ -0,0 +1,99 @@
+ * @param <E> The class type of the {@link Estimator} implementation itself
+ * @param <M> class type of the {@link Model} this Estimator produces.
+ */
+public abstract class Estimator<E extends Estimator <E, M>, M extends Model <M>>
```
Thanks, changed to EstimatorBase.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/Estimator.java:
```diff
@@ -0,0 +1,99 @@
+import org.apache.flink.table.api.java.StreamTableEnvironment;
+
+/**
+ * Abstract class for a estimator that fit a {@link Model}.
```
Thanks, changed.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/Estimator.java:
```diff
@@ -0,0 +1,99 @@
+	 * @param input the table with records to train the Model.
+	 * @return a model trained to fit on the given Table.
+	 */
+	public M fit(Table input) {
```
It is better to be public. We have refined the JavaDoc.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/Estimator.java:
```diff
@@ -0,0 +1,99 @@
+	 */
+	@Override
+	public M fit(TableEnvironment tEnv, Table input) {
+		MLSession.setTableEnvironment(tEnv, input);
```
Thanks for your advice and offline discussion, we defined MLEnvironmentFactory to replace the global static MLSession.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/AlgoOperator.java:
```diff
@@ -0,0 +1,98 @@
+ * Base class for algorithm operators.
+ * @param <T> The class type of the {@link AlgoOperator} implementation itself
+ */
+public abstract class AlgoOperator<T extends AlgoOperator <T>> implements WithParams<T>, Serializable {
```
Thanks, we have add more description. 
PipelineStage only supports single input and single output, it is the basic unit for pipeline. AlgoOperator supports multi-input and multi-output. We’d like to implement the algorithm with AlgoOperator, and PipelineStage’s fit and transform function can call the AlgoOperator.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/AlgoOperator.java:
```diff
@@ -0,0 +1,98 @@
+		}
+	}
+
+	public static AlgoOperator sourceFrom(Table table) {
```
thx, added

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLSession.java:
```diff
@@ -0,0 +1,187 @@
+	private static StreamTableEnvironment streamTableEnv;
+
+	/**
+	 * Factory to create {@link ExecutionEnvironment}.
```
thx, changed

xuyang1706 commented:
> @xuyang1706 Thanks for the patch. I left some comments. In general it would avoid class name collision. We may also want to consider if the packaging can change a little bit. For example:
> 
> ```
> org.apache.flink.ml.
>   .algooperator
>       .AlgoOperator.java
>       .stream
>           .StreamAlgoOperator.java
>           .source
>                .TableSourceStreamAlgoOperator.java
>       .batch
>           .BatchAlgoOperator.java
>           .source
>                TableSourceBatchAlgoOperator.java
> ```

@becketqin , thanks for your advice. we defined MLEnvironmentFactory to replace the global static MLSession and support to set multiple Environment, refactored the class names and refined the javaDoc.

xuyang1706 commented:
> Thanks for the contribution @xuyang1706 and sorry for joining the discussion late. Please see some of my comments. In general I think there are lots of added APIs that we should carefully document. Please let me know what you guys think ;-) thanks -Rong

@walterddr , thanks for your comments. We added more description on the concepts and APIs, and refactored the core functions of link, linkTo and linkFrom.

ex00 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/source/TableSourceBatchOp.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public final class TableSourceBatchOp extends BatchOperator<TableSourceBatchOp> {
+
+	public TableSourceBatchOp(Table table) {
```
this logic could be moved to parent, it same as in `TableSourceStreamOp `

xuyang1706 replied:
refer to https://github.com/apache/flink/pull/9184#discussion_r326942376

ex00 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/AlgoOperator.java:
```diff
@@ -0,0 +1,156 @@
+	@Override
+	public Params getParams() {
+		if (null == this.params) {
+			this.params = new Params();
```
duplicate logic from constructor. 
If we cant set `params` as null we don't need check any time it.

xuyang1706 replied:
Thanks for your advice, removed the check process.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/AlgoOperator.java:
```diff
@@ -0,0 +1,156 @@
+	 */
+	public static AlgoOperator<?> sourceFrom(Table table) {
+		if (((TableImpl) table).getTableEnvironment() instanceof StreamTableEnvironment) {
+			return new TableSourceStreamOp(table);
```
Thanks, removed.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/AlgoOperator.java:
```diff
@@ -0,0 +1,156 @@
+	/**
+	 * Construct the operator with empty Params.
+	 */
+	protected AlgoOperator() {
```
The Table is the output of the operator. It should be set at linkFrom and call getOutput to get the output for the next linked operator in most of the scenes. So the AlgoOperator() and the AlgoOperator(Params params) should be exist.
The TableSourceBatchOp and the TableSourceStreamOp is the only two cases that construct the Operator using the table. These operators are the special cases and need not add a new constructor in the base class for these cases.

walterddr replied:
I think the main point here is that since the `public AlgoOperator()`  API is not used/tested. we can add it in from later PRs, correct me if I were wrong @ex00 ? 
If that's the case then yes I do agree that any new APIs added which are not tested should be part of latter PRs.

ex00 replied:
yes @walterddr, you are right. I don't see how using this constructor here.
Also, personally for me, logic in parent class, when using `output` field and it is not marked 'as required', is confusing a little bit.

xuyang1706 replied:
In later PRs, these APIs will be used and tested.

For how using this constructor, I'd like to give an example. 
SplitBatchOp is widely used in ML data pre-preprossing, which splits one dataset input 2 dataset: training set and validation set. It is very convenient for us to write code like this:
new SplitBatchOp().setSplitRatio(0.9)

For output field. The AlgoOperator may have one or more result tables, in most cases, it has only one result. The output is the main operation result table, and the other results are kept in the sideOutputs.
For example, 2 AlgoOperators: AlgoA and AlgoB, AlgoB takes the AlgoA’s results as its inputs, we can write the code like this:
AlgoB.linkFrom(AlgoA)
AlgoA.getOutput() provides the main result of AlgoA, and AlgoA.getSideOutputs() provides the other results of AlgoA. AlgoB will take the AlgoA’s results as its inputs by calling AlgoA.getOutput() and AlgoA.getSideOutputs().

ex00 replied:
>new SplitBatchOp().setSplitRatio(0.9)

How will it be apply to data source and how will define output result? Could you explain more please?

xuyang1706 replied:
Thanks, we have the `setOutput` and can set the output in any method of the operator

ex00 replied:
is it mean what it will be look like `new SplitBatchOp().setSplitRatio(0.9).setOutput(table)` ?
if it is right assumption, that mean that who will write code need to know that need to call `setOutput`. And if someone will doing something from you example, easier to lost this call, and exception will be throw in run time only
```
AlgoA = new SplitBatchOp().setSplitRatio(0.9)
AlgoB.linkFrom(AlgoA)
AlgoA.getSchema()  // NPE: output is null
```
But if we define that `SplitBatchOp(table)` and developer don't thing about additional method calls and will get error on compile time in wrong case.

xuyang1706 replied:
Yes, there must be one or more data sources. We have defined some common *SourceBatchOp to get the source data conveniently, and from the sourceBatchOp, we can link the algorithm operations.
Here is a whole example:
```
CsvSourceBatchOp algoA =
	new CsvSourceBatchOp()
		.setFilePath("http://alink-dataset.cn-hangzhou.oss.aliyun-inc.com/csv/iris.csv")
		.setSchemaStr("s_length double, s_width double, p_length double, p_width double, category string");

System.out.println("schema of algoA:");
System.out.println(algoA.getSchema());

SplitBatchOp algoB = new SplitBatchOp().setFraction(0.8);

algoB.linkFrom(algoA);

System.out.println("schema of algoB's main result:");
System.out.println(algoB.getSchema());

System.out.println("schema of algoB's side output result:");
System.out.println(algoB.getSideOutput().getSchema());
```

And the print results:
```
schema of algoA:
root
 |-- s_length: Double
 |-- s_width: Double
 |-- p_length: Double
 |-- p_width: Double
 |-- category: String

schema of algoB's main result:
root
 |-- s_length: Double
 |-- s_width: Double
 |-- p_length: Double
 |-- p_width: Double
 |-- category: String

schema of algoB's side output result:
root
 |-- s_length: Double
 |-- s_width: Double
 |-- p_length: Double
 |-- p_width: Double
 |-- category: String
```

ex00 replied:
now I see, thanks for the explanation

xuyang1706 commented:
> Hi @xuyang1706 thanks for update,
> I've left additional comments, please look they.

@ex00 , thanks for your advice,  I have refactored the sourceFrom() method and removed the unnecessary check process.

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,127 @@
+ *
+ * <p>This class is extended to support the data transmission between the BatchOperators.
+ */
+public abstract class BatchOperator<T extends BatchOperator<T>> extends AlgoOperator<T> {
```
one suggestions I have is to put the operators into a single package
for example
```
|--operator
   |--AlgoOperator.java
   |--stream
      |--StreamOperator.java
      |--source
         |--TableSourceStreamOperator.java
   |--batch
      |--BatchOperator.java
      |--source
         |--TableSourceBatchOperator.java
```
IMO this is more clear than putting `AlgoOperator` in common. and others scatter around the package directory, what do you think?

xuyang1706 replied:
Thanks @becketqin , @walterddr , @ex00 , I have refactored the package paths according your suggestions.

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironment.java:
```diff
@@ -0,0 +1,178 @@
+ * @see StreamTableEnvironment
+ */
+public class MLEnvironment {
+	private ExecutionEnvironment env;
```
any reason MLEnvironment supports both stream and batch at the same time? or is this just to avoid having to split into `StreamMLEnv` and `BatchMlEnv` which both implements some sort of common interface?

xuyang1706 replied:
Yes, it is intended to be this for simplicity.

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironment.java:
```diff
@@ -0,0 +1,178 @@
+					this.streamEnv = ((StreamTableEnvironmentImpl) streamTableEnv).execEnv();
+				}
+			}
+		} else {
```
I would still put a check here 
```
else if (tEnv instanceof BatchTableEnvironment) {
  //...
} else {
  throw new IllegalArgumentEception(...);
}
```

xuyang1706 replied:
Thanks for your advice, added.

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/EstimatorBase.java:
```diff
@@ -0,0 +1,101 @@
+ * @param <E> The class type of the {@link EstimatorBase} implementation itself
+ * @param <M> class type of the {@link ModelBase} this Estimator produces.
+ */
+public abstract class EstimatorBase<E extends EstimatorBase<E, M>, M extends ModelBase<M>>
```
I might've been wrong, but looks like `EstimatorBase` is not tested?

walterddr replied:
same as other Pipeline stages.

xuyang1706 replied:
Thanks, we have add some cases, and we will add more cases with the extended class.

xuyang1706 commented:
> Thanks for the follow up @xuyang1706 . Overall the changes looks good to me. I left some minor comments. Also I notices that some large trunk of code was missing tests? were we going to support individual tests in the future, or they are part of the `MLSessionTest`?

Thanks @walterddr , I refactored the package paths, added some test cases and refined the JavaDoc with examples.

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/AlgoOperator.java:
```diff
@@ -0,0 +1,163 @@
+
+import java.io.Serializable;
+
+/**
```
The class java doc here is a little confusing, especially given that there is no `linkFrom()` method in `AlgoOperator` class. My understanding is the following:


Base class for the algorithm operators. It hosts the parameters and output tables of an algorithm operator. Each AlgoOperator may have one or more output tables. One of the output table is the primary output table which can be obtained by calling {@link #getOutput}. The other output tables are side output tables that can be obtained by calling {@link #getSideOutput()}.

The input of an AlgoOperator is defined in the subclasses of the AlgoOperator.

xuyang1706 replied:
Thanks, changed

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/AlgoOperator.java:
```diff
@@ -0,0 +1,163 @@
+	/**
+	 * Construct the operator with empty Params.
+	 *
+	 * <p>For how using this constructor, there is an example:
```
This constructor is especially useful when users want to set parameters for the algorithm operators. For example, .....

xuyang1706 replied:
Thanks, changed

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/AlgoOperator.java:
```diff
@@ -0,0 +1,163 @@
+ *
+ * @param <T> The class type of the {@link AlgoOperator} implementation itself
+ */
+public abstract class AlgoOperator<T extends AlgoOperator <T>>
```
Extra space before <T>

xuyang1706 replied:
Thanks, changed

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/stream/StreamOperator.java:
```diff
@@ -0,0 +1,130 @@
+ *
+ * <p>This class is extended to support the data transmission between the StreamOperators.
+ */
+public abstract class StreamOperator<T extends StreamOperator<T>> extends AlgoOperator<T> {
```
Can we change the java doc in this class to match those in BatchOperator?

xuyang1706 replied:
Thanks, changed

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/stream/StreamOperator.java:
```diff
@@ -0,0 +1,130 @@
+		return new TableSourceStreamOp(table);
+	}
+
+	protected void checkOpSize(int size, StreamOperator<?>... inputs) {
```
Why these methods are not static here just like they are in BatchOperator? Also, it seems that these methods can be merged with the methods in BatchOperator by changing the parameter type to AlgoOperator.

xuyang1706 replied:
Thanks, these methods in the BatchOperator is also not static. The BatchOperator and the StreamOperator should not be mixed in the array

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/EstimatorBase.java:
```diff
@@ -0,0 +1,101 @@
+	 * @param input the table with records to train the Model.
+	 * @return a model trained to fit on the given Table.
+	 */
+	public M fit(Table input) {
```
Does this method have to be public?

xuyang1706 replied:
Thanks, this method is more convenient with the MLEnvironmentFactory, so I think it should be public

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/PipelineStageBase.java:
```diff
@@ -0,0 +1,66 @@
+import org.apache.flink.ml.params.shared.HasMLEnvironmentId;
+
+/**
+ * The base class for a stage in a pipeline, either an [[Estimator]] or a [[Transformer]].
```
It seems the java doc format for the classes are not quite consistent. Sometimes it is `Class`, sometimes <code>Class</code>, and here it is [[Class]]. Can we make them consistent?

xuyang1706 replied:
Thanks, changed

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/PipelineStageBase.java:
```diff
@@ -0,0 +1,66 @@
+/**
+ * The base class for a stage in a pipeline, either an [[Estimator]] or a [[Transformer]].
+ *
+ * <p>Each pipeline stage is with parameters, and requires a public empty constructor for
```
The PipelineStageBase maintains the parameters for the stage. A default constructor is needed in order to restore a pipeline stage.

xuyang1706 replied:
Thanks, changed

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/TransformerBase.java:
```diff
@@ -0,0 +1,100 @@
+/**
+ * The base class for transformer implementations.
+ *
+ * @param <T> The class type of the {@link TransformerBase} implementation itself, used by {@link
```
The class type of the {@link TransformerBase} implementation itself =>

A subclass of {@link TransformerBase}, used by ...

xuyang1706 replied:
Thanks, changed

becketqin commented on flink-ml-parent/flink-ml-lib/src/test/java/org/apache/flink/ml/common/MLEnvironmentTest.java:
```diff
@@ -0,0 +1,87 @@
+/**
+ * Test cases for MLEnvironment.
+ */
+public class MLEnvironmentTest {
```
It looks that there is no assertion in this test class.

xuyang1706 replied:
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,136 @@
+/**
+ * Base class of batch algorithm operators.
+ *
+ * <p>This class is extended to support the data transmission between the BatchOperators.
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,136 @@
+	/**
+	 * Abbreviation of {@link #linkTo(BatchOperator)}.
+	 */
+	public <B extends BatchOperator<?>> B link(B next) {
```
Thanks, removed the linkTo

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,127 @@
+	 * }
+	 * </pre>
+	 *
+	 * <p>the <code>c</code> in upper code is the linked
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,127 @@
+	 * <p>the <code>c</code> in upper code is the linked
+	 * <code>b</code> which use <code>a</code> as input.
+	 *
+	 * @param next the linked BatchOperator
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,127 @@
+	/**
+	 * Link to another {@link BatchOperator}.
+	 *
+	 * <p>Link the <code>next</code> to BatchOperator using this as its input.
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,127 @@
+	 * }
+	 * </pre>
+	 *
+	 * <p>the <code>d</code> in upper code is the linked
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,127 @@
+	 * <p>the <code>d</code> in upper code is the linked
+	 * <code>c</code> which use a and b as its inputs.
+	 *
+	 * <p>note: It is not recommended to linkFrom itself or linkFrom the same group inputs twice.
```
Thanks, the subclass define the behavior, so we add the tips in the comment only.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,127 @@
+	 * @param table the input table
+	 * @return the new BatchOperator
+	 */
+	public static BatchOperator<?> sourceFrom(Table table) {
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/batchoperator/BatchOperator.java:
```diff
@@ -0,0 +1,127 @@
+			+ size + ", current: " + inputs.length);
+	}
+
+	protected void checkRequiredOpSize(int size, BatchOperator<?>... inputs) {
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironment.java:
```diff
@@ -0,0 +1,184 @@
+import org.apache.flink.util.Preconditions;
+
+/**
+ * MLEnvironment hold the execution environment.
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironment.java:
```diff
@@ -0,0 +1,184 @@
+	 *
+	 * @param env the ExecutionEnvironment
+	 */
+	public void setExecutionEnvironment(ExecutionEnvironment env) {
```
Thanks, removed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironment.java:
```diff
@@ -0,0 +1,184 @@
+	 *
+	 * @param env the StreamExecutionEnvironment
+	 */
+	public void setStreamExecutionEnvironment(StreamExecutionEnvironment env) {
```
Thanks, removed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironment.java:
```diff
@@ -0,0 +1,184 @@
+	 * @param tEnv  the TableEnvironment
+	 * @param table the table for get {@link ExecutionEnvironment} for batch
+	 */
+	public void setTableEnvironment(TableEnvironment tEnv, Table table) {
```
Thanks, removed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironment.java:
```diff
@@ -0,0 +1,184 @@
+	 * @see MLEnvironment#setExecutionEnvironment(ExecutionEnvironment)
+	 */
+	public ExecutionEnvironment getExecutionEnvironment() {
+		if (null == env) {
```
Thanks, I have been added the Constructor that use the environment as its parameters

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironmentFactory.java:
```diff
@@ -0,0 +1,98 @@
+	/**
+	 * Map that hold the MLEnvironment and use the MLEnvironmentId as its key.
+	 */
+	private static HashMap<Long, MLEnvironment> map = new HashMap<>();
```
Thanks, added

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironmentFactory.java:
```diff
@@ -0,0 +1,98 @@
+ * Factory to get the MLEnvironment using a MLEnvironmentId.
+ *
+ * <p>The MLEnvironmentFactory manages the multiple MLEnvironments by the MLEnvironmentId.
+ * Two steps that the user use the Factory:
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironmentFactory.java:
```diff
@@ -0,0 +1,98 @@
+		return map.get(mlEnvId);
+	}
+
+	static {
```
Thanks, removed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironmentFactory.java:
```diff
@@ -0,0 +1,98 @@
+	 *
+	 * @return the MLEnvironment id.
+	 */
+	public static synchronized Long getNewMLEnvironmentId() {
```
Thanks, added

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/AlgoOperator.java:
```diff
@@ -0,0 +1,163 @@
+	/**
+	 * Returns the column types of the output table.
+	 */
+	public TypeInformation <?>[] getColTypes() {
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/batch/BatchOperator.java:
```diff
@@ -0,0 +1,127 @@
+	 * }
+	 * </pre>
+	 *
+	 * <p>the <code>c</code> in upper code is the linked
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/batch/BatchOperator.java:
```diff
@@ -0,0 +1,127 @@
+	 * }
+	 * </pre>
+	 *
+	 * <p>the <code>d</code> in upper code is the linked
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/batch/BatchOperator.java:
```diff
@@ -0,0 +1,127 @@
+	 * @param table the input table
+	 * @return the new BatchOperator
+	 */
+	public static BatchOperator<?> sourceFrom(Table table) {
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/batch/BatchOperator.java:
```diff
@@ -0,0 +1,127 @@
+			+ size + ", current: " + inputs.length);
+	}
+
+	protected void checkRequiredOpSize(int size, BatchOperator<?>... inputs) {
```
Thanks, changed

xuyang1706 commented:
> @xuyang1706 Thanks for updating the patch. I left a few more comments. Can you take a look?

Thanks, the comments in the code have been refined. The `sourceFrom` has been renamed to fromTable. The constructor with the Environment has been added in the MLEnvironment. And the register

Thanks @becketqin, I have refined the code according your comments.

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/TransformerBase.java:
```diff
@@ -0,0 +1,98 @@
+	}
+
+	@Override
+	public Table transform(TableEnvironment tEnv, Table input) {
```
Should we do a sanity check to make sure the tEnv passed in is the same table environment used by the input table?

xuyang1706 replied:
I agree with replace `void fit(TableEnvironment, Table)` method with the `MLEnvironmentFactory`, and I suggest we could discuss and change API in another PR.

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/pipeline/EstimatorBase.java:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	@Override
+	public M fit(TableEnvironment tEnv, Table input) {
```
Should we do a sanity check to make sure the tEnv passed in is the same one used by the input table?

xuyang1706 replied:
Ditto above.

becketqin commented on flink-ml-parent/flink-ml-lib/src/test/java/org/apache/flink/ml/common/MLEnvironmentTest.java:
```diff
@@ -0,0 +1,64 @@
+		Assert.assertSame(mlEnvironment.getStreamExecutionEnvironment(), streamExecutionEnvironment);
+		Assert.assertSame(mlEnvironment.getStreamTableEnvironment(), streamTableEnvironment);
+	}
+}
```
If we only expect the MLEnvironment to be constructed in this two ways, maybe we can just have those two specific constructors?

xuyang1706 replied:
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironmentFactory.java:
```diff
@@ -0,0 +1,125 @@
+	public static final Long DEFAULT_ML_ENVIRONMENT_ID = 0L;
+
+	/**
+	 * A 'id' is a unique identifier of a MLEnvironment.
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/AlgoOperator.java:
```diff
@@ -0,0 +1,153 @@
+	/**
+	 * Returns the column names of the output table.
+	 */
+	public String[] getColNames() {
```
Thanks, added

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/stream/StreamOperator.java:
```diff
@@ -0,0 +1,134 @@
+			+ size + ", current: " + inputs.length);
+	}
+
+	protected void checkMinOpSize(int size, StreamOperator<?>... inputs) {
```
Thanks, changed

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironment.java:
```diff
@@ -0,0 +1,130 @@
+	 */
+	public StreamExecutionEnvironment getStreamExecutionEnvironment() {
+		if (null == streamEnv) {
+			streamEnv = StreamExecutionEnvironment.getExecutionEnvironment();
```
Thanks, as we discussed offline, for depending an unfixed bug, we will use current implementation as a temp workaround.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironmentFactory.java:
```diff
@@ -0,0 +1,125 @@
+	 *
+	 * @param env the MLEnvironment
+	 */
+	public static synchronized void setDefault(MLEnvironment env) {
```
Thanks, I have checked the existence of default MLEnv, user could only set one default MLEnv.

xuyang1706 commented:
> @xuyang1706 Thanks for updating the patch. I left a few more comments. Most of them are minor. One major thing is I am not sure whether we should completely remove `fit(TableEnvironment, Table)` and `transform(TableEnvironment, Table)` API, because they are completely bypassed in the current implementation given the introduction of `MLEnvironmentFactory`.

Thanks, @becketqin . I agree with replace `void fit(TableEnvironment, Table)` method with the `MLEnvironmentFactory`, and I suggest we could discuss and change the API in another PR.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironmentFactory.java:
```diff
@@ -0,0 +1,125 @@
+			if (mlEnvId.equals(DEFAULT_ML_ENVIRONMENT_ID)) {
+				setDefault(new MLEnvironment());
+			} else {
+				throw new IllegalArgumentException("There is no Environment in factory. " +
```
Thanks for your advice, we refactored code to create the default ML Session in the static block and changed the exception message.

becketqin commented:
@xuyang1706 Thanks for updating the patch. the patch LGTM overall. I created a PR with some minor improvements against your branch. Can you check if that makes sense? If so, feel free to merge it to your branch.

@walterddr Do you also want to take another look? I am thinking of merging the patch either on Sunday or Monday. Thanks.

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/MLEnvironmentFactory.java:
```diff
@@ -0,0 +1,111 @@
+	 * @param mlEnvId the id.
+	 * @return the removed MLEnvironment
+	 */
+	public static synchronized MLEnvironment remove(Long mlEnvId) {
```
when would this API be called? seems like it is not necessary?

xuyang1706 replied:
MLEnviromentFactory can create multi MLEnviroment, if some of them only used in a time slot, user can use this API to release the resource. In most cases, user just use default MLEnviroment. This API is rarely used.

becketqin replied:
There is one minor bug here. The default MLEnvironment should never be removed. I'll fix this when checking in the code. Thanks.

xuyang1706 commented:
> Thanks for the contribution @xuyang1706 and the reviews @becketqin . Overall it looks good to me. I only left a minor comment. please kindly take a look.
> +1 to merge to unblock future developments.

Thanks @walterddr. Yes, this API is rarely used, but in some cases, if user want to release no-used MLEnviroment in the process, this is only API could be called. Thus, I prefer to keep it.

xuyang1706 commented:
> @xuyang1706 Thanks for updating the patch. the patch LGTM overall. I created a PR with some minor improvements against your branch. Can you check if that makes sense? If so, feel free to merge it to your branch.
> 
> @walterddr Do you also want to take another look? I am thinking of merging the patch either on Sunday or Monday. Thanks.

Thanks for your help @becketqin. I have merged your "minor improvement PR" to my branch.

becketqin commented:
Merged to master.
